### PR TITLE
SIL: remove the SILOpenedArchetypesTracker

### PR DESCRIPTION
**Pull request by eeckstein**
Instead, put the archetype->instruction map into SILModule.

SILOpenedArchetypesTracker tried to maintain and reconstruct the mapping locally, e.g. during a use of SILBuilder.
Having a "global" map in SILModule makes the whole logic _much_ simpler. I'm wondering why we didn't do this in the first place.

It's now required that opened archetypes must be unique in a module - which makes sense. This was the case anyway, except for keypath accessors and in some sil test files.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -3521,6 +3521,14 @@ SILGenModule::emitKeyPathComponentForDecl(SILLocation loc,
     return storage->isSettable(storage->getDeclContext());
   };
 
+  // We cannot use the same opened archetype in the getter and setter. Therefore
+  // we create a new one for the setter.
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp:3525]: Doesn't the getter also share an opened archetype with the original expression?

**Comment by slavapestov** [comment]
This is really cool, nice work!

File: lib/SILGen/SILGenExpr.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SILGen/SILGenExpr.cpp:3525]: What I've seen is that the opened archetype is only used in the getter and the setter. So making a new one for the setter is sufficient

Comment (eeckstein) [reply lib/SILGen/SILGenExpr.cpp:3525]: On the other hand, it does not harm to create new opened archetypes for the getter as well

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Pull request merged**